### PR TITLE
feature: implemented new lease-related api

### LINKED ISSUE #65: lease supported of etcd v3
To better control time-to-live of keys, one could bind multiple keys to a single lease.

### PR DESCRIPTION
**Pull request by Yiyiyimu**
Fix #65 

1. implement lease-related api
2. add test file
3. add documentation

also fixed a naming typo

### CODE REVIEW

**Comment by membphis** [comment]
@Yiyiyimu please take a look at the output of Travis-CI .

**Comment by Yiyiyimu** [comment]
Hi @membphis I'm still working on the errors of other etcd versions. Is there anyway to set it to WIP? Or I just need to convert it to draft?

Also to deal with test in multi-version etcd locally, I'm trying to build dockerfile to automate the test process. Do you think that's necessary? Or that's not a big deal for others that building docker on my own is enough?

**Comment by membphis** [comment]
@Yiyiyimu I have updated the title, you can continue your job now.

you can remove `WIP` after you complete this feature.

**Comment by Yiyiyimu** [comment]
Thank you @membphis ! Do you have any suggestions on test of multi-version etcd locally?

**Comment by membphis** [comment]
> Thank you @membphis ! Do you have any suggestions on test of multi-version etcd locally?

we can run the specified etcd at our local machine, and try to find the bug. you can take a look at the travis script, it is useful.

**Comment by Yiyiyimu** [comment]
Fix two problems:
1. the different API of lease-related function in multi-version
2. leases() is only supported after v3.4

Actually leases() is supported in v3.3.12, as [etcd documentation](https://etcd.io/docs/v3.3.12/dev-guide/api_reference_v3/#service-lease-etcdserveretcdserverpbrpcproto) shows. Since currently we deploy v3.3.0, I test it only in v3.4.0 and add another file `lease3.4.t`.

Considering this kind of problem could be common in the future, and this kind of naming is lack of readability, is there any need to build different branches in this repo according to different versions of etcd?

**Comment by Yiyiyimu** [comment]
> we can run the specified etcd at our local machine, and try to find the bug. you can take a look at the travis script, it is useful.

Thank you @membphis it works!

File: t/v3/lease.t [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+            local res, err = etcd:grant(2)
+            check_res(res, err)
+
+            local data, err = etcd:set("/test", "abc", {prev_kv = true, lease = res.body.ID})
+            check_res(data, err)
+
+            local data, err = etcd:get("/test")
+            check_res(data, err, "abc")
+
+            ngx.sleep(2.5)
```
[the remark this answers is not in the document]
Comment (Yiyiyimu) [reply t/v3/lease.t:71]: fixed

File: t/v3/lease.t [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+--- config
+    location /t {
+        content_by_lua_block {
+            local etcd, err = require "resty.etcd".new({protocol = "v3"})
+            check_res(etcd, err)
+
+            local res, err = etcd:grant(5)
+            check_res(res, err)
+
+            local data, err = etcd:set("/test", "abc", {prev_kv = true, lease = res.body.ID})
```
[the remark this answers is not in the document]
Comment (Yiyiyimu) [reply t/v3/lease.t:100]: Fixed. Test on attaching multiple keys to one lease, and using revoke to delete them all.

File: t/v3/lease.t [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
```
[the remark this answers is not in the document]
Comment (Yiyiyimu) [reply t/v3/lease.t:136]: Fixed. Test on separately attaching two keys to two leases, keepalive one and to see if one alive and another one expired.

File: t/v3/lease.t [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+            check_res(data, err)
+
+            local data, err = etcd:get("/test")
+            check_res(data, err, "abc")
+
+            local data, err = etcd:timetolive(res.body.ID)
+            check_res(data, err)
+
+            ngx.sleep(1)
+            if tonumber(data.body.TTL) < 5 then
```
[the remark this answers is not in the document]
Comment (Yiyiyimu) [reply t/v3/lease.t:146]: It's my fault. I add sleep after the comparison but forget to delete the comparison.

Also, I changed testing logic here from read TTL parameter to get key after default expired time.

File: t/v3/lease3.4.t [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+            local res, err = etcd:grant(2)
+            check_res(res, err)
+
+            local data, err = etcd:set("/test", "abc", {prev_kv = true, lease = res.body.ID})
+            check_res(data, err)
+
+            local data, err = etcd:get("/test")
+            check_res(data, err, "abc")
+
```
[the remark this answers is not in the document]
Comment (Yiyiyimu) [reply t/v3/lease3.4.t:75]: Fixed.

**Comment by Yiyiyimu** [comment]
Improve test file in:
1. multiple leases
2. multiple keys conbined to one lease	
3. the keys responced from timetolive are the same to granted (add decoding)
4. manually assign ID when grant lease

**Comment by Yiyiyimu** [comment]
> It seems one v2 test file which I seems didn't change failed, I'll have a look on it.

Same with #12

**Comment by Yiyiyimu** [comment]
Do we need to add the response structure of each API in doc, like only the body part, as mentioned in #37 

For example `timetolive`
```
# +  body = {
# +    ID = "1",
# +    TTL = "2",
# +    grantedTTL = "3",
# +    header = {
# +      cluster_id = "14841639068965178418",
# +      member_id = "10276657743932975437",
# +      raft_term = "2",
# +      revision = "199"
# +    },
# +    keys = { "/test1" }
# +  },
```

and `get`
```
# +  body = {
# +    count = "1",
# +    header = {
# +      cluster_id = "14841639068965178418",
# +      member_id = "10276657743932975437",
# +      raft_term = "2",
# +      revision = "184"
# +    },
# +    kvs = { {
# +        create_revision = "184",
# +        key = "/test2",
# +        lease = "1",
# +        mod_revision = "184",
# +        value = "bcd",
# +        version = "1"
# +      } }
# +  },
```

**Comment by nic-chen** [comment]
it is better to add

**Comment by Yiyiyimu** [comment]
> it is better to add

Sure I'll create another PR for this change.

**Comment by membphis** [comment]
@Yiyiyimu many thx

**Pull request merged**